src/storage/transaction/TransactionUtils.cpp

```diff
@@ -0,0 +1,124 @@
+        << "\n ver = " << NebulaKeyUtils::getVersion(vIdLen, rawKey)
+        << "\n size() = " << rawKey.size() << "\n hex=\n"
+        << folly::hexDump(rawKey.data(), rawKey.size());
+    return oss.str();
+}
+
+std::string TransactionUtils::dumpEdgeKey(const cpp2::EdgeKey& key) {
+    std::ostringstream oss;
+    oss << "edgekey: src " << key.src << ", type=" << key.edge_type << ", ranking=" << key.ranking
+        << ", dst=" << key.dst << ". ";
+    return oss.str();
+}
+
+std::string TransactionUtils::dumpNewEdge(const cpp2::NewEdge& edge) {
+    std::ostringstream oss;
+    auto strEdgeKey = dumpEdgeKey(edge.key);
+        oss << strEdgeKey << "\n" << folly::hexlify(strEdgeKey) << "\n";
+    for (auto& value : edge.props) {
+        oss << value.toString() << " ";
+    }
+    return oss.str();
+}
+
+std::string TransactionUtils::dumpAddEdgesRequest(const cpp2::AddEdgesRequest& req) {
+    std::ostringstream oss;
+    oss << "\nAddEdgesRequest: "
+        << ", space_id=" << req.space_id << "\n"
+        << ", parts.size()=" << req.parts.size();
+    for (auto& part : req.parts) {
+        oss << "\npart=" << part.first;
+        for (auto& edge : part.second) {
+            oss << ", edge=" << dumpNewEdge(edge) << "\n";
+        }
+    }
+    oss << "prop_names.size()=" << req.prop_names.size();
+    oss << "\nprops_name=";
+    for (auto& prop : req.prop_names) {
+        oss << prop << " ";
+    }
+    return oss.str();
+}
+
+std::string TransactionUtils::edgeKey(size_t vIdLen,
+                                      PartitionID partId,
+                                      const cpp2::EdgeKey& key,
+                                      EdgeVersion ver) {
+    return NebulaKeyUtils::edgeKey(
+        vIdLen, partId, key.src.getStr(), key.edge_type, key.ranking, key.dst.getStr(), ver);
+}
+
+cpp2::EdgeKey TransactionUtils::reverseEdgeKey(const cpp2::EdgeKey& key) {
+    cpp2::EdgeKey ret{key};
+    std::swap(ret.src, ret.dst);
+    ret.edge_type = 0 - ret.edge_type;
+    return ret;
+}
+
+cpp2::NewEdge TransactionUtils::reverseEdge(const cpp2::NewEdge& edge) {
+    cpp2::NewEdge ret(edge);
+    std::swap(ret.key.src, ret.key.dst);
+    ret.key.edge_type = 0 - ret.key.edge_type;
+    return ret;
+}
+
+std::string TransactionUtils::reverseRawKey(size_t vIdLen,
+                                            PartitionID partId,
+                                            const std::string& rawKey) {
+    return NebulaKeyUtils::edgeKey(vIdLen,
+                                   partId,
+                                   NebulaKeyUtils::getSrcId(vIdLen, rawKey).str(),
+                                   NebulaKeyUtils::getEdgeType(vIdLen, rawKey),
+                                   NebulaKeyUtils::getRank(vIdLen, rawKey),
+                                   NebulaKeyUtils::getDstId(vIdLen, rawKey).str(),
+                                   NebulaKeyUtils::getVersion(vIdLen, rawKey));
```

Comment:
There is no exchange of srcID and dstID here?

Reply:
good catch & fixed.